### DISCUSSION TITLE
Testing Environments

### DISCUSSION THREAD

**Discussion by djcyr** [bloxbean/cardano-client-lib#156] [Q&A] [answered]
While using branch 0.3.0-beta2, I was expecting to have something available along the lines of:

Networks.devnet();
Networks.preview();
Networks.preproduction();

But still only see the two: 

Networks.testnet();   
Networks.mainnet();

While attempting to use the Network class constructor with magic numbers of 9, 2, and 1 respectively, received exceptions.

Is it too soon to be able to connect?

**Comment by satran004** [comment]
Ideally you should be able to create your own instance of Network class and use it.

Can you please explain what exactly you are trying with some sample code (if possible)? Which backend provider are you using ? 

But I agree, we should add other public networks to the Networks class. I will create an enhancement for it.

Fyi.
Network Id in Network is used for address generation and it can have only two values 1 (magic) or 0 (for testnets including dev, preview, prepod, or custom)

Network magic/protocolMagic in Network is not directly used but  has different values for different networks.

Thanks

**Reply by djcyr** [reply]
...was simply trying to get a balance for an address from preview testnet.
		 
I intuited that I would instantiate the network class as 
				new Network(0b0000,2);
but it barked at me with the following exception:		 
AddressRuntimeException: Unknown network type - Network{network_id=0, protocol_magic=2}

(I use cardano-cli query tip --testnet-magic 2 to successfully query the node. 
I thought the second argument to the network constructor was "magic" ?) 		
		 		 
As it turns out, since all the testnets have a common address prefix of addr_test1,
it seems I can get away with using Networks.testnet() as the argument when constructing 
the account.   I was able to successfully query a balance with the following code: 


	String mnumonic = "one two .. twentyfour";
	Account account = new Account(Networks.testnet(), mnumonic);

	BFBackendService backendService = 
		new BFBackendService("https://cardano-preview.blockfrost.io/api/v0/",api_key);
			
	Result<AddressContent> result = 
		backendService.getAddressService().getAddressInfo(account.baseAddress());
			
	if (result.isSuccessful() {
			
		List<TxContentOutputAmount> amount = backendService.getAddressService()
				.getAddressInfo(account.baseAddress()).getValue().getAmount();
			
		... amount.size());
		... amount.get(0).getQuantity();	
	}

Would there be any problem later if I create a signed transaction 
with that account having the wrong magic number (value 2 instead of 1097911063)?

Thanks!

**Comment by satran004** [comment, marked as answer]
You are right.  There are only two possible values for network Id,  0 (all testnets) and 1 (mainnet).
NetworkId is used during address generation, not protocol magic.

 But somewhere in the code, the whole network object is compared during address generation to find the address prefix. This is a bug. We need to fix that to check only networkId in the Network object.
 
Yes, the workaround is - always pass Networks.testnet() during non mainnet address generation. 
 
 This will be fixed as part of this PR to avoid any confusion.
 https://github.com/bloxbean/cardano-client-lib/pull/158 
 
 >> Would there be any problem later if I create a signed transaction
with that account having the wrong magic number (value 2 instead of 1097911063)?

No. It should work. The transaction object has an optional parameter "networkId" for either mainnet / testnet transaction. 
Protocol magic is only used during the communication between client (Blockfrost) and node. AFAIK it doesn't have any role in transaction serialization.
